File: frontend/lib/features/printers/domain/printer.dart
// ignore_for_file: invalid_annotation_target
import 'package:freezed_annotation/freezed_annotation.dart';
import 'package:json_annotation/json_annotation.dart';

part 'printer.freezed.dart';
part 'printer.g.dart';

@JsonEnum(alwaysCreate: true)
enum PrinterTechnology {
  @JsonValue('FDM')
  fdm,
  @JsonValue('SLA')
  sla,
}

@JsonEnum(alwaysCreate: true)
enum PrinterConnectorType {
  @JsonValue('octoprint')
  octoprint,
  @JsonValue('prusalink')
  prusalink,
  @JsonValue('mock')
  mock,
  @JsonValue('manual')
  manual,
}

@JsonEnum(alwaysCreate: true)
enum PrinterStatusValue {
  @JsonValue('idle')
  idle,
  @JsonValue('printing')
  printing,
  @JsonValue('error')
  error,
  @JsonValue('offline')
  offline,
  @JsonValue('maintenance')
  maintenance,
}

extension PrinterTechnologyX on PrinterTechnology {
  String get apiValue {
    switch (this) {
      case PrinterTechnology.fdm:
        return 'FDM';
      case PrinterTechnology.sla:
        return 'SLA';
    }
  }
}

extension PrinterStatusValueX on PrinterStatusValue {
  String get apiValue {
    switch (this) {
      case PrinterStatusValue.idle:
        return 'idle';
      case PrinterStatusValue.printing:
        return 'printing';
      case PrinterStatusValue.error:
        return 'error';
      case PrinterStatusValue.offline:
        return 'offline';
      case PrinterStatusValue.maintenance:
        return 'maintenance';
    }
  }
}

@freezed
class Printer with _$Printer {
  const factory Printer({
    required String id,
    required String name,
    String? model,
    required PrinterTechnology technology,
    @JsonKey(name: 'build_volume_x') double? buildVolumeX,
    @JsonKey(name: 'build_volume_y') double? buildVolumeY,
    @JsonKey(name: 'build_volume_z') double? buildVolumeZ,
    @JsonKey(name: 'connector_type')
    required PrinterConnectorType connectorType,
    required PrinterStatusValue status,
    @JsonKey(name: 'materials_supported') List<String>? materialsSupported,
    @JsonKey(name: 'last_seen_at') DateTime? lastSeenAt,
    @JsonKey(name: 'created_at') required DateTime createdAt,
  }) = _Printer;

  factory Printer.fromJson(Map<String, dynamic> json) =>
      _$PrinterFromJson(json);
}

@freezed
class PrinterCreate with _$PrinterCreate {
  const factory PrinterCreate({
    required String name,
    String? model,
    required PrinterTechnology technology,
    @JsonKey(name: 'build_volume_x') double? buildVolumeX,
    @JsonKey(name: 'build_volume_y') double? buildVolumeY,
    @JsonKey(name: 'build_volume_z') double? buildVolumeZ,
    @JsonKey(name: 'connector_type')
    @Default(PrinterConnectorType.mock)
    PrinterConnectorType connectorType,
    @JsonKey(name: 'connection_url') String? connectionUrl,
    @Default(PrinterStatusValue.offline) PrinterStatusValue status,
    @JsonKey(name: 'materials_supported') List<String>? materialsSupported,
    @JsonKey(name: 'api_key') String? apiKey,
  }) = _PrinterCreate;

  factory PrinterCreate.fromJson(Map<String, dynamic> json) =>
      _$PrinterCreateFromJson(json);
}

@freezed
class PrinterUpdate with _$PrinterUpdate {
  @JsonSerializable(includeIfNull: false)
  const factory PrinterUpdate({
    String? name,
    String? model,
    PrinterTechnology? technology,
    @JsonKey(name: 'build_volume_x') double? buildVolumeX,
    @JsonKey(name: 'build_volume_y') double? buildVolumeY,
    @JsonKey(name: 'build_volume_z') double? buildVolumeZ,
    @JsonKey(name: 'connector_type') PrinterConnectorType? connectorType,
    @JsonKey(name: 'connection_url') String? connectionUrl,
    PrinterStatusValue? status,
    @JsonKey(name: 'materials_supported') List<String>? materialsSupported,
    @JsonKey(name: 'api_key') String? apiKey,
  }) = _PrinterUpdate;

  factory PrinterUpdate.fromJson(Map<String, dynamic> json) =>
      _$PrinterUpdateFromJson(json);
}

File: backend/app/schemas/printer.py
"""
Pydantic v2 schemas for Printer.

Note: PrinterRead intentionally omits `api_key_encrypted`. Never add it.
A test in tests/test_printers.py asserts this invariant.
"""

from datetime import datetime
from typing import Literal, Optional
from uuid import UUID

from pydantic import BaseModel, Field

# Reusable enum-like Literal types
Technology = Literal["FDM", "SLA"]
ConnectorType = Literal["octoprint", "prusalink", "mock", "manual"]
PrinterStatusValue = Literal["idle", "printing", "error", "offline", "maintenance"]


# ---------- Base / Create / Update ----------


class PrinterBase(BaseModel):
    """Fields shared between create and update payloads."""

    name: str
    model: Optional[str] = None
    technology: Technology
    build_volume_x: Optional[float] = None
    build_volume_y: Optional[float] = None
    build_volume_z: Optional[float] = None
    connector_type: ConnectorType = "mock"
    connection_url: Optional[str] = None
    materials_supported: Optional[list[str]] = None


class PrinterCreate(PrinterBase):
    # Plaintext API key — encrypted by the service before persisting.
    api_key: Optional[str] = Field(default=None, repr=False)
    status: PrinterStatusValue = "offline"


class PrinterUpdate(BaseModel):
    """All fields optional — partial update."""

    name: Optional[str] = None
    model: Optional[str] = None
    technology: Optional[Technology] = None
    build_volume_x: Optional[float] = None
    build_volume_y: Optional[float] = None
    build_volume_z: Optional[float] = None
    connector_type: Optional[ConnectorType] = None
    connection_url: Optional[str] = None
    materials_supported: Optional[list[str]] = None
    status: Optional[PrinterStatusValue] = None
    # Plaintext — re-encrypted by the service if provided
    api_key: Optional[str] = Field(default=None, repr=False)


# ---------- Read (response model) ----------


class PrinterRead(BaseModel):
    """Public response shape. NEVER includes api_key_encrypted."""

    id: UUID
    name: str
    model: Optional[str] = None
    technology: Technology
    build_volume_x: Optional[float] = None
    build_volume_y: Optional[float] = None
    build_volume_z: Optional[float] = None
    connector_type: ConnectorType
    status: PrinterStatusValue
    materials_supported: Optional[list[str]] = None
    last_seen_at: Optional[datetime] = None
    created_at: datetime

    model_config = {"from_attributes": True}


# ---------- Status endpoint ----------


class PrinterStatus(BaseModel):
    """Shape returned by GET /printers/{id}/status."""

    printer_id: UUID
    status: PrinterStatusValue
    current_job_id: Optional[UUID] = None
    progress_pct: Optional[float] = None
    temperature_nozzle: Optional[float] = None
    temperature_bed: Optional[float] = None
    last_seen_at: Optional[datetime] = None

    model_config = {"from_attributes": True}


# ---------- Test-connection endpoint ----------


class PrinterTestResult(BaseModel):
    """Shape returned by POST /printers/{id}/test."""

    printer_id: UUID
    ok: bool
    message: str